### ISSUE TITLE
Lightbulb performance improvements

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn#66968]
> cc: @genlu and @akhera99. This would also benefit from better observability in the Lightbulb space and in beefing up dashboards.

@arkalyanms @genlu @akhera99 I am also seeing quite a good lightbulb performance improvement with this PR. I have done both OutOfProc and InProc measurements for both the Main branch and PR branch, using the exact same repro steps. 
1. **OutOfProc:** On main branch the lightbulb progress bar goes ~4-5 times before the quick actions get populated, while on this PR branch, the quick actions get populated after the progress bar goes around ~2-2.5 times. Approximately 2X improvement.
2. **InProc:** On main branch the lightbulb progress bar goes ~2 times before the quick actions get populated, while on this PR branch, the quick actions get populated even before the progress goes around once. Approximately 2X-3X improvement.

**Summary:** It seems there is a 2X improvement in lightbulb perf for this scenario from this PR, both for InProc and OutOfProc analysis, and there is scope for further 2X-3X improvement after this PR goes in by moving just lightbulb diagnostics to InProc. Once this PR goes in, I will experiment moving the lightbulb diagnostic computation to always run InProc, while leaving the background analysis to be OutOfProc/InProc based on the above Tools Option.

### Main branch ==> Performance in Current shipping VS
**4-5 progress bar cycles**

![OutOfProc_Main_Branch_LightBulb_Perf](https://user-images.githubusercontent.com/10605811/219355727-b381482d-c629-43b0-93b7-bf8b8fa6e739.gif)

### PR branch ==> Performance after this PR is merged
**2-2.5 progress bar cycles**

![OutOfProc_PR_Branch_LightBulb_Perf](https://user-images.githubusercontent.com/10605811/219357115-fdc64374-b032-4a43-8825-4f80ac838adc.gif)

### Main branch
**2 progress bar cycles**

![InProc_Main_Branch_LightBulb_Perf](https://user-images.githubusercontent.com/10605811/219355808-93fa54f7-cfb6-46c4-823f-efbeee38494f.gif)

### PR branch ==> Desired default performance in VS
**< 1 progress bar cycle**

![InProc_PR_Branch_LightBulb_Perf](https://user-images.githubusercontent.com/10605811/219355834-7c268a4f-ba25-451a-948c-412bd40f8a35.gif)

_Originally posted by @mavasani in https://github.com/dotnet/roslyn/issues/66750#issuecomment-1432833465_

**Comment by CyrusNajmabadi** [comment]
> Once this PR goes in, I will experiment moving the lightbulb diagnostic computation to always run InProc,

This doesn't make sense, and it not a viable path forward.  We must be able to get the *best* experience possible, including hosts that will only be able to run us OOP, or through systems like LSP.

There is nothnig about in-proc/oop that should add any perf overhead for lightbulbs.  In proc should literally only be presentation of a few strings in a lightbulb.  There should be practically no perf impact involved in that.  If we are seeing perf slowness in OOP for these scenarios, then something very wrong is going on and we need to address that (especially as that's necessary for all our hosts).

This is not speculative either.  I've been through this several times with other features (like Classification, FAR, Rename, etc.) where some features were doing things poorly (sometimes even splitting their work in the different hosts).  Ensuring the feature properly does all it's work in OOP and only uses the host for UI has always produced a result where OOP vs inproc perf are equivalent.  That should be what we have with lightbulbs as well, and if not we have to figure out why and solve that part.

**Comment by mavasani** [comment]
Tagging @CyrusNajmabadi @sharwell 

I did a bunch of measurements and experiments to try and understand the difference between InProc and OOP performance numbers for the lightbulb.

1. I first tried to capture perfview traces for both of them to see any outliers, but that did not give me anything useful. The only unrelated thing I identified is that we always compute all the `PreviewWorkspace` diagnostics InProc. This is due to the below if condition being true for the workspace not being a `VisualStudioWorkspace`. However, this did not turn out to affect the lightbulb population time (as expected). https://github.com/dotnet/roslyn/blob/4190056e0126f206c64439275fbf43a54dd31067/src/VisualStudio/Core/Def/Remote/VisualStudioRemoteHostClientProvider.cs#L56-L66

2. I then experimented by tweaking the code in the core method in `CompilationWithAnalyzers` that computes analyzer diagnostics: https://github.com/dotnet/roslyn/blob/4190056e0126f206c64439275fbf43a54dd31067/src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs#L572-L614

3. To confirm whether the OOP messaging/serialization/solution syncing is adding observable overhead to the scenario, I first tweaked the method to be a no-op (just return at start). With this change, both OOP and InProc lightbulb population became almost instantaneous and there were no observable execution time differences. This confirms @CyrusNajmabadi's claim that the performance overhead is not coming from the OOP-InProc syncing. 

4. Subsequently, I moved the early bailout in the method to after the 3rd step in the implementation note below, i.e. after cloning the compilation with the new event queue, initializing the analyzer driver and executing all the compilation start actions to get all registered analyzer actions. This did not add any noticeable overhead to measurement in step 3 above for either InProc or OOP. https://github.com/dotnet/roslyn/blob/4190056e0126f206c64439275fbf43a54dd31067/src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs#L590-L639 

5. Then, I moved the return to just after the first half of the 4th step, i.e. invocation below to `GetCompilationEventsForSingleFileAnalysis`. This method basically executes `SemanticModel.GetDiagnostics` call for the file. Note that for analyzers that can compute span based diagnostics, we execute `SemanticModel.GetDiagnostics(span)`, but because we always have full document analyzers, such as remove unused usings, that need to run on the entire file to report diagnostics on a span, we do end up invoking `SemanticModel.GetDiagnostics(fullFileSpan)`. This lead to a noticeable overhead in the OOP computation experience, it added more than half a progress bar cycle to the async lightbulb population, while added a slit second in the InProc computation experience (about ~10% overhead). https://github.com/dotnet/roslyn/blob/4190056e0126f206c64439275fbf43a54dd31067/src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs#L667-L669

6. Then, I moved the return statement to after the below if statement for SymbolStart/End analyzers, i.e. force compute compilation events for all the files with partial declarations for a symbol declared in the original file. For my repro, this type has bunch of partials, so this basically leads to `SemanticModel.GetDiagnostics(fullFileSpan)` invocations for all these files. This adds about an entire progress bar cycle for OOP computation and half a cycle for InProc computation. https://github.com/dotnet/roslyn/blob/4190056e0126f206c64439275fbf43a54dd31067/src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs#L667-L701

7. Note that after step 6 above, the only remaining step in this method is the core analyzer callbacks based on the generated compilation events. I finally moved the return to after the below line to also make callbacks, and this adds a minor but equivalent overhead in both OOP and InProc computation experience. https://github.com/dotnet/roslyn/blob/4190056e0126f206c64439275fbf43a54dd31067/src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs#L701

## Summary

- The core overhead between OOP and InProc computation seems to be coming from `SemanticModel.GetDiagnostics()` calls. The same calls on same set of files are just slowed when executed in OOP. I am still unable to figure out the underlying reason, I'll keep investigating, but wanted to point it out upfront in case someone from the IDE team has a theory on it.

- The overhead coming from SymbolStart/End analyzers needing to run on all partial types seems to add a significant overhead. So I played around with moving all the SymbolStart/End analyzers and corresponding code fixes to a separate `CodeActionPriorityRequest.Low` bucket for async lightbulb, and this gives significant performance improvements as we populate all the code fixes from non-SymbolStart/End analyzers and code refactorings in about half a progress bar cycle for OOP, and almost instantaneously for InProc. The async lightbulb does take the required additional one cycle to then populate the fixes for SymbolStart/End analyzers and the Suppress/Configure actions, but that seems completely reasonable to me. We only have a handful of SymbolStart/End analyzers and it seems reasonable to me to move them below to improve the overall lightbulb performance. I am going to open a separate PR with gifs showing the lightbulb experience before/after this change. I would recommend looking at this PR and screenshots before evaluating this performance optimization and experience.

- I will create a separate tracking item to see if PreviewWorkspace diagnostic computation can also be done in the same process as other diagnostics computation. I also noticed that this diagnostic computation is not hitting the inproc diagnostic cache (likely due to version/workspace differences), but we should try to workaround these so populating the preview can be made faster.

**Comment by mavasani** [comment]
## Summary for Lightbulb Performance improvements

**Test file:** https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

**Lines of Code:** ~13.2K

**Conclusions:**
   1. From 17.5.3 to 17.6.Preview2
      1. ~2X to 3X improvement in lightbulb performance
      2. Enabling ServerGC leads to another ~2X performance improvement over No ServerGC on 17.5.3 as well as on 17.6.Preview2 

   2. From 17.6.Preview2 to 17.7.1 IntPreview latest dogfood build
      1. Another ~2X-3X improvement in lightbulb performance
      2. Less than 0.5 progress bar cycles, almost instantaneous lightbulb in the file
      3. Performance is almost instantaneous for both ServerGC enabled and not enabled.
      4. We don't require enabling ServerGC anymore to get the best lightbulb performance in 17.7.1 IntPreview as the most expensive compute is avoided/delayed now.

### 1) 17.5.3

#### a) No ServerGC
![17_5_3_Lightbulb](https://user-images.githubusercontent.com/10605811/229683054-df8b1c1a-f4c0-47ac-823a-3745f50222e7.gif)

#### b) ServerGC
![17_5_3_Lightbulb](https://user-images.githubusercontent.com/10605811/229683323-e8b51c94-85b0-40cf-8211-49fa6887fa6c.gif)


### 2) 17.6.Preview2

#### a) No ServerGC
![17_6_Preview2_Lightbulb](https://user-images.githubusercontent.com/10605811/229683107-5df41d18-114e-40a4-a298-38443a28015a.gif)

#### b) ServerGC
![17_6_Preview2_Lightbulb](https://user-images.githubusercontent.com/10605811/229683339-ea440152-f67a-4fc9-855e-331ea379c63e.gif)


### 3) 17.7.1 Latest IntPreview Dogfood build

#### a) No ServerGC
![17_7_IntPreview_Lightbulb](https://user-images.githubusercontent.com/10605811/229683245-833031a1-0df7-4fad-ba6e-d6ac4c4f9852.gif)

#### b) ServerGC
![17_7_IntPreview_Lightbulb](https://user-images.githubusercontent.com/10605811/229683358-207fc657-1fc0-49c6-84dd-9406892407fb.gif)
